actionpack/lib/action_dispatch/http/uri.rb

```diff
@@ -0,0 +1,247 @@
+        parts = host.split(".")
+        parts[0..-(tld_length + 2)]
+      end
+
+      def build_host_url(host, port, protocol, options, path)
+        if match = host.match(HOST_REGEXP)
+          protocol ||= match[1] unless protocol == false
+          host       = match[2]
+          port       = match[3] unless options.key? :port
+        end
+
+        protocol = normalize_protocol protocol
+        host     = normalize_host(host, options)
+
+        result = protocol.dup
+
+        if options[:user] && options[:password]
+          result << "#{Rack::Utils.escape(options[:user])}:#{Rack::Utils.escape(options[:password])}@"
+        end
+
+        result << host
+        normalize_port(port, protocol) { |normalized_port|
+          result << ":#{normalized_port}"
+        }
+
+        result.concat path
+      end
+
+      def named_host?(host)
+        !IP_HOST_REGEXP.match?(host)
+      end
+
+      def normalize_protocol(protocol)
+        case protocol
+        when nil
+          secure_protocol ? "https://" : "http://"
+        when false, "//"
+          "//"
+        when PROTOCOL_REGEXP
+          "#{$1}://"
+        else
+          raise ArgumentError, "Invalid :protocol option: #{protocol.inspect}"
+        end
+      end
+
+      def normalize_host(_host, options)
+        return _host unless named_host?(_host)
+
+        tld_length = options[:tld_length] || @@tld_length
+        subdomain  = options.fetch :subdomain, true
+        domain     = options[:domain]
+
+        host = +""
+        if subdomain == true
+          return _host if domain.nil?
+
+          host << extract_subdomains_from(_host, tld_length).join(".")
+        elsif subdomain
+          host << subdomain.to_param
+        end
+        host << "." unless host.empty?
+        host << (domain || extract_domain_from(_host, tld_length))
+        host
+      end
+
+      def normalize_port(port, protocol)
+        return unless port
+
+        case protocol
+        when "//" then yield port
+        when "https://"
+          yield port unless port.to_i == 443
+        else
+          yield port unless port.to_i == 80
+        end
+      end
+  end
+
+  def initialize(uri_string)
+    if ipv6 = uri_string.match(IP6_REGEX)
+      @uri = ::URI::Generic.build2(scheme: ipv6[1], host: ipv6[2], port: ipv6[3], path: ipv6[4])
+    else
+      # get_header("HTTP_HOST") || "#{server_name}:#{get_header('SERVER_PORT')}"
+      @uri = ::URI.parse(uri_string)
+    end
+  end
+
+  def protocol
+    "#{scheme}://"
+  end
+
+  # Returns a \host:\port string for this request, such as "example.com" or
+  # "example.com:8080". Port is only included if it is not a default port
+  # (80 or 443).
+  def host_with_port
+    "#{host}#{port_string}"
+  end
+
+  # Returns a string \port suffix, including colon, like ":8080" if the \port
+  # number of this request is not the default HTTP \port 80 or HTTPS \port 443.
+  def port_string
+    standard_port? ? "" : ":#{port}"
+  end
```

Comment:
Most of the others are more plausible, but this doesn't seem public-method-worthy to me. I'd rather not encourage callers to reconstruct a URL via string-bashing.

Reply:
yes, sounds about right. Happy to change that. 
I believe there is good potential to refactor parts of the class and instance methods. And I started with it, but I reverted most of it, after feedback to do this in a 2nd step as it might be used by others and changing the API might break other people's code..